src/ucs/sys/topo.c

```diff
@@ -11,6 +11,40 @@
 #include <ucs/sys/topo.h>
 #include <ucs/type/status.h>
 #include <stdio.h>
+#include <ucs/datastruct/khash.h>
+#include <ucs/type/spinlock.h>
+#include <ucs/debug/log.h>
+#include <ucs/debug/assert.h>
+
+#define UCS_TOPO_MAX_SYS_DEVICES 1024
+
+typedef int64_t ucs_bus_id_bit_rep_t;
+
+typedef enum ucs_topo_kh_put {
+    UCS_KH_PUT_FAILED = -1,
+    UCS_KH_PUT_KEY_PRESENT,
+    UCS_KH_PUT_BUCKET_EMPTY,
+    UCS_KH_PUT_BUCKET_CLEAR
+} ucs_topo_kh_put_t;
+
+typedef struct ucs_topo_sys_dev_to_bus_arr {
+    ucs_sys_bus_id_t bus_arr[UCS_TOPO_MAX_SYS_DEVICES];
+    unsigned count;
+} ucs_topo_sys_dev_to_bus_arr_t;
+
+KHASH_INIT(bus_to_sys_dev /*name*/,
+           ucs_bus_id_bit_rep_t /*key*/,
+           ucs_sys_device_t /*val*/,
+           1 /*is_map*/,
+           kh_int64_hash_func, kh_int64_hash_equal)
+
+static khash_t(bus_to_sys_dev) ucs_topo_bus_to_sys_dev_hash;
+static ucs_recursive_spinlock_t  ucs_topo_bus_to_sys_dev_lock;
+
+static ucs_topo_sys_dev_to_bus_arr_t ucs_topo_sys_dev_to_bus_lookup;
+static ucs_recursive_spinlock_t  ucs_topo_sys_dev_to_bus_lock;
+
+static int ucs_topo_initialized = 0;
```

Comment:
let's add wrapper struct: ucs_topo_global_context
and pull all gobal vars in it

Reply:
done